traversal/fns.go

```diff
@@ -67,3 +81,17 @@ type SkipMe struct{}
 func (SkipMe) Error() string {
 	return "skip"
 }
+
+type ErrBudgetExceeded struct {
+	BudgetKind string // "node"|"link"
+	Path       datamodel.Path
+	Link       datamodel.Link // only present if BudgetKind=="link"
+}
+
+func (e *ErrBudgetExceeded) Error() string {
+	msg := fmt.Sprintf("traversal budget exceeded: budget for %ss reached zero as we reached path %q", e.BudgetKind, e.Path)
```

Comment:
Perhaps
```suggestion
	msg := fmt.Sprintf("traversal budget exceeded: budget for %ss reached zero while on path %q", e.BudgetKind, e.Path)
```

The rest of the PR looks 👍